### DISCUSSION TITLE
If I want to scale the convex hull of a rigid body, what is the correct operation?

### DISCUSSION THREAD

**Discussion by CeSun** [notgiven688/jitterphysics2#28] [How do I ... ?]
Transform all vertices and reconstruct the shape?

**Comment by notgiven688** [comment]
> Transform all vertices and reconstruct the shape?

Yes, transforming all vertices and reconstructing the shape would be the method of choice. However, there's an exception: if you have many instances of convex hulls that only differ in their scaling factor. In such cases, you can wrap your `ConvexHullShape` inside a `TransformedShape`. To perform the scaling transformation, modify the diagonal entries of the "orientation" matrix in the constructor. I will rename the matrix, as it supports not only orientation but also scaling. In fact, all affine transformations (see [Affine Transformation](https://en.wikipedia.org/wiki/Affine_transformation)) are supported by `TransformedShape`.

Here's an example from JitterDemo, in ConvexDecomposition.cs:

Change:

```cs
foreach (ConvexHullShape s in shapesToAdd)
{
  body.AddShape(s.Clone(), false);
}
```

To:

```cs
foreach (ConvexHullShape s in shapesToAdd)
{
  body.AddShape(new TransformedShape(s.Clone(), JMatrix.Identity * 0.2f, JVector.Zero), false);
}
```

This will scale your object by a factor of 0.2.

**Comment by CeSun** [comment]
> modify the diagonal entries of the "orientation" matrix in the constructor

Orientation does not seem to take effect on scaling,  my code is:

```c#
RigidBody = Owner.CurrentLevel.PhyWorld.CreateRigidBody();
RigidBody.Position = new JVector(WorldLocation.X, WorldLocation.Y, WorldLocation.Z);
foreach(var shape in StaticMesh.Shapes)
{
    RigidBody.AddShape(shape);
}
var Matrix = MatrixHelper.CreateTransform(Vector3.Zero, WorldRotation, WorldScale);
RigidBody.Orientation = new JMatrix
{
    M11 = Matrix.M11,
    M12 = Matrix.M12,
    M13 = Matrix.M13,
    M21 = Matrix.M21,
    M22 = Matrix.M22,
    M23 = Matrix.M23,
    M31 = Matrix.M31,
    M32 = Matrix.M32,
    M33 = Matrix.M33,
};
```

**Comment by CeSun** [comment]
Modifying the shape every time the scale changes is very resource intensive, and will also generate a lot of objects that generate GC pressure, so I hope to directly modify the matrix of the rigid body.

**Comment by notgiven688** [comment]
You need to wrap your hull-shape in a TransformedShape - see my example from above. You can then change the transforms "on the fly" without much cost, by modifying the "Transformation" property of the TransformedShape.

**Reply by CeSun** [reply]
Good job! at first, I thought TransformedShape could only transform the shape of the input parameter during construction...

**Reply by notgiven688** [reply]
I found a small bug in the code: When you set the transformation matrix via the `Transformation` property, all mass properties (geometric center, mass, inertia) of the `TransformedShape` get recalculated by recalculating the mass properties of the child shape (in your case a `PointCloudShape`). For a `PointCloudShape` this can be rather slow. https://github.com/notgiven688/jitterphysics2/pull/42 fixes this problem.
